_posts/2023-10-04-Data-Quality-Checks.md

```diff
@@ -0,0 +1,408 @@
+# Example of using the jsonvalidate R package.
+# Referenced with modifications from:
+# https://docs.ropensci.org/jsonvalidate/articles/jsonvalidate.html
+
+schema <- '{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "title": "Hello World JSON Schema",
+  "description": "An example",
+  "type": "object",
+  "properties": {
+    "hello": {
+      "description": "Provide a description of the property here",
+      "type": "string"
+    }
+  },
+  "required": [
+    "hello"
+  ]
+}'
+
+# create a schema contract for data
+validate <- jsonvalidate::json_validator(schema, engine = "ajv")
+
+# validate JSON using schema specification contract and invalid data
+validate("{}")
+
+# validate JSON using schema specification contract and valid data
+validate("{'hello':'world'}")
+```
+
+[JSON Schema](https://json-schema.org/learn/getting-started-step-by-step) provides a vocabulary way to validate schema contracts for JSON documents.
+There are several implementations of the vocabulary, including [Python package jsonschema](https://github.com/python-jsonschema/jsonschema), and R package [jsonvalidate](https://github.com/ropensci/jsonvalidate).
+Using these libraries allows you to define pre- or postcondition data schema contracts for your software work.
+See above for an R based example of using this vocabulary to perform data schema testing.
+
+## Shift-left Data Testing
+
+```mermaid!
+flowchart LR
+    subgraph local ["Data Workflow as Hoare Triple"]
+        direction LR
+            subgraph spacer [" "]
+                subgraph question ["Data source condition?"]
+                    input_data[("Input Data\n(P - precondition)")]
+                end
+            end
+            process_data["Data processing\n(C - command)"]
+            output_data[("Output Data\n(Q - postcondition)")]
+    end
+
+    input_data --> process_data --> output_data
+
+    style question fill:#ffffff,stroke:#ffffff;
+    style spacer fill:#ffffff
+```
+
+Earlier portions of this article have covered primarily data validation of command side-effects and postconditions.
+This is commonplace in development where data sources usually are provided without the ability to validate their precondition or definition.
+[Shift-left testing](https://en.wikipedia.org/wiki/Shift-left_testing) is a movement which focuses on validating earlier in the lifecycle if and when possible to avoid downstream issues which might occur.
+
+### Shift-left Data Testing - Data Version Control (DVC)
+
+```mermaid!
+flowchart LR
+
+    dvc[("Data Version\nControl (DVC)")]
+    versioned_data[("Versioned\nDatasets")]
+    dvc_cli["DVC Client"]
+    input_data[("Input Data\n(P - precondition)")]
+
+    dvc --> | version data | dvc_cli
+    versioned_data --> | actual data | dvc_cli
+    dvc_cli --> | to maintain | input_data
+```
+
+Data sources undergoing frequent changes becomes difficult to use because we oftentimes don't know _when_ the data is from or what version it might be.
```

Comment:
```suggestion
Data sources undergoing frequent changes become difficult to use because we oftentimes don't know _when_ the data is from or what version it might be.
```

Reply:
Thank you! Adding this in.